src/Orleans/Lifecycle/LifecycleObservable.cs

```diff
@@ -0,0 +1,125 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Orleans.Runtime;
+
+namespace Orleans
+{
+    public class LifecycleObservable<TStage> : ILifecycleObservable<TStage>, ILifecycleObserver
```

Comment:
Type classes is something people were discussing the language repo a while ago. Not sure it will hit somewhere someday.